sources/api/shibaken/src/warmpool/autoscaling_warm_pool.rs

```diff
@@ -0,0 +1,132 @@
+/// This module contains utilities for waiting until the instance on warm pool to be
+/// marked as InService.
+use crate::warmpool::error::{self, Result};
+use argh::FromArgs;
+use imdsclient::ImdsClient;
+use log::{info, warn};
+use serde::Deserialize;
+use snafu::{OptionExt, ResultExt};
+use std::fs;
+use std::path::Path;
+use tokio::time::{sleep, Duration};
+
+// Path to config file containing host's autoscaling.should-wait setting
+const CONFIG_PATH: &str = "/etc/warm-pool-wait.toml";
+const FETCH_LIFECYCLE_INTERVAL_SECS: Duration = Duration::from_secs(5);
+
+#[derive(FromArgs, Debug)]
+#[argh(subcommand, name = "warm-pool-wait")]
+/// Fetch to IMDS and wait until instance is marked as InService.
+pub(crate) struct WarmPoolWait {}
+
+impl WarmPoolWait {
+    pub(crate) async fn run(self) -> Result<()> {
+        let config_parse = get_config().await?;
+
+        let should_wait_value = config_parse.should_wait;
+        let marker_file_path = config_parse.marker_path;
+
+        info!("autoscaling.should-wait value is {}", should_wait_value);
+        if should_wait_value {
+            wait_until_inservice().await?;
+        }
+
+        info!("Marker file path is {}", marker_file_path);
+        fs::write(&marker_file_path, "").unwrap_or_else(|e| {
+            warn!(
+                "Failed to create marker file '{}', shibaken service may unexpectedly run again: '{}'",
+                &marker_file_path, e
+            )
+        });
+        Ok(())
+    }
+}
+
+/// Uses an imdsclient function to fetch the lifecyclestate of the host
+async fn get_lifecycle_state(client: &mut ImdsClient) -> Result<String> {
+    client
+        .fetch_autoscaling_lifecycle_state()
+        .await
+        .context(error::ImdsRequestSnafu)?
+        .context(error::ImdsNoneSnafu {
+            what: "instance-id",
+        })
+}
+
+/// Continuously fetches lifecycle state of the host.
+/// Returns from function when host is "InService"
+async fn wait_until_inservice() -> Result<()> {
+    let mut client = ImdsClient::new();
+
+    loop {
+        let lifecycle_state = get_lifecycle_state(&mut client).await?;
+
+        if lifecycle_state.eq("InService") {
+            info!("Lifecycle state is {} ... exiting.", lifecycle_state);
+            break;
+        } else {
+            info!("Lifecycle state is {} ... waiting.", lifecycle_state);
+            sleep(FETCH_LIFECYCLE_INTERVAL_SECS).await;
```

Comment:
Doesn't this return a `Future` that must be consumed like `await?` as in the line above?

Reply:
Not need here due to `the ? operator can only be applied to values that implement Try`. I believe it's safe here and we have [example](https://github.com/bottlerocket-os/bottlerocket/blob/67e5abfb0e1575fb6bcb7a710cdb3d422ef12283/sources/api/apiclient/src/update.rs#L252)